orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -958,6 +962,64 @@ IMPLEMENT_GRADIENT_BUILDER(GetGeluGradient) {
               {GI(0)})};
 }
 
+namespace {
+std::vector<NodeDef> GetBiasGeluGradNodes(
+    bool use_approximation,
+    const ArgDef& dY, const ArgDef& X, const ArgDef& B,  // inputs
+    const ArgDef& X_plus_B,                              // intermediate arguments
+    const ArgDef& dX, const ArgDef& dB) {                // outputs
+  ORT_ENFORCE(GetShape(B).size() == 1, "B must have exactly one dimension.");
+
+  // reduce all but last dimension of dX
+  const std::vector<int64_t> dX_reduction_axes = [&dX]() {
+    const auto dX_num_dims = GetShape(dX).size();
+    std::vector<int64_t> result(dX_num_dims - 1);
+    std::iota(result.begin(), result.end(), int64_t{0});
+    return result;
+  }();
+
+  return std::vector<NodeDef>{
+      NodeDef("Add",
+              {X, B},
+              {X_plus_B}),
+      NodeDef(OpDef{use_approximation ? "FastGeluGrad" : "GeluGrad", kMSDomain, 1},
+              {dY, X_plus_B},
+              {dX}),
```

Comment:
The next step is to fuse these two nodes together ...

We probably still need an op like 

BiasGeluGradPartial and same for FastGeluGradPartial

Reply:
This addition would be free when fused into the following kernel 

---
In reply to: [436212709](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r436212709) [](ancestors = 436212709)